ls1intum/Artemis
# [Feature] Re-evaluate feedback and result after structured grading instructions updated

### Checklist

### Motivation and Context


This PR is second part of #3294 . Basically, it aims to re-evaluate the exercises when the instructor updates the structured grading instructions if these instructions are  already associated with feedbacks during assessment period. 

### Description

To achieve this re-evaluation, the new "/re-evaluate" endpoint is introduced for each type of exercises. Currently, when there are changes (e.g. updating the credit/deleting instruction or criterion) for grading instructions are associated with feedbacks, Artemis warns the user with popup before updating the exercise. If the user decides to click "Confirm Changes", this action will only update the exercise and the grading instructions.  With this PR, there is additional "Re-evaluate" button. So, the result will be re-calculated and the feedbacks which is associated with grading instructions will updated. 

Since, each type of exercise has own resource, to trigger the `updateXExercise`, `../re-evaluate` endpoint is added for each exercise. After re-evaluation is completed, the usual update is running for the corresponding exercise.  

Additionally, user can decide to remove all related feedbacks if he/she delete grading instruction is used to assess submissions. For that, `exercise-update-warning` popup displays `deleteFeedback` checkbox. If user choses this option, "Confirm Changes" button will be de-activated automatically. Because, the results need to be re-calculated. After that, client sends `deleteFeedback`  as optional request parameter. If it sets as `true`, the grading instructions of exercise is retrieved from database as a backup. After collecting each deleted instruction in one list, it finds all feedbacks are associated with deleted grading instructions, then it removes all of them. After that, it re-calculates the results. Re-evaluation is running before update process, because to find feedbacks will be deleted, I need the existing exercise grading instructions. 

**Please look at the translation files in more detail. Cause, my German is really bad and I'm using deepl for translation, so would be great if you find the translation problems 😄 

### Steps for Testing


To be sure everything is working correctly, the testing steps should be followed each type of exercise except quiz. (Cause structured grading instructions are not relevant with quizzes).  

Side not: If you want to test this for the programming exercise, you should activate the `manual assessment`. First you should set the `due date` for the exercise on timeline, then activate/click `Manual Review`. After that, you will see the `Structured Grading Instruction switch button` on the edit page.

1. Log in to Artemis
2. Navigate to Course Administration
3. Create/Find a new exercise.
4. Add an example solution for the exercise / participate this exercise as a student.
 a. If you add example submission, then create new assessment for this submission and use one of the structured grading instruction as a feedback.
 b. If you participated this exercise as student, then login as tutor/instructor and assess this submission. During the assessment, use one of the structured grading instructions as a feedback.
5. After that, as an instructor, go to edit view for this exercise.
6. switch the structured grading instruction button, after first usage of the feedback, the instructions should be displayed in form.
7. Try following actions:
 a. Delete grading criterion which you used the sub-grading instruction to assess submission at the previous steps
 b. Delete instruction which you used to assess submission at the previous steps
 c. Change `credit` of grading instruction which you used to assess submission at the previous steps
 d. Change `maxCountInScore` of grading instruction which you used to assess submission at the previous steps
8. You should see the warning pop-up. Then try following steps (a and b) independently: 
 a. Do not click "delete all feedbacks" checkbox (If you can not see this checkbox, you probably did not delete grading criterion/instruction), just click "Re-evaluate" button. 
     I. If you open the assessment again, check the score of feedback. It should be updated.
    II. Check the total score, it should be updated.
 b. click "delete all feedbacks" checkbox, "confirm changes" button should be disabled, then click "Re-evaluate" button.
    I. If you open the assessment again, the feedback should not be there.
   II. Check the total score, it should be updated.
 

### Test Coverage



ProgrammingExerciseResource.java: 91%
TextExerciseResource,java:  91%
ModelingExerciseResource,java:  87% (reEvaluate covered only)
FileUploadExerciseResource,java:  100%
ExerciseService,java:  90%
exercise-update-warning.component.ts: 100%


### Screenshots


<img src="https://user-images.githubusercontent.com/23650630/121421151-26d4d500-c96e-11eb-8dd0-39d4c185af6a.png" width="400" />
